bst.py

```diff
@@ -145,6 +153,47 @@ def _breadth_first_helper(self, node):
             if root.right_child:
                 q.append(root.right_child)
 
+    def delete(self, val):
+        """Remove a node from the tree if it exists, connect child nodes to new
+        parents"""
+        node = self.contains(val)
+        try:
+            self.size -= 1
+            # found node.
+            # if node has no kids, remove node:
+            if node.right_child is None and node.left_child is None:
+                if node.is_left:
+                    node.parent.left_child = None
+                elif node.is_right:
+                    node.parent.right_child = None
+                return
+            else:
+                swap_node = self._find_second_largest(node)
+                node.val = swap_node.val
+                if swap_node.is_left:
+                    if swap_node.left_child:
+                        swap_node.parent.left_child = swap_node.left_child
+                    else:
+                        swap_node.parent.left_child = swap_node.right_child
+                elif swap_node.is_right:
+                    if swap_node.left_child:
+                        swap_node.parent.right_child = swap_node.left_child
+                    else:
+                        swap_node.parent.right_child = swap_node.right_child
```

Comment:
I think you're trying to collapse what are really two separate cases into one.  I don't think it's going to work well for you.